src/widget/textbox.rs

```diff
@@ -14,114 +14,120 @@
 };
 
-const ACTIVE_BORDER_COLOR: Color = Color::rgb24(0xff_00_00);
-const INACTIVE_BORDER_COLOR: Color = Color::rgb24(0x55_55_55);
+const BACKGROUND_GREY_LIGHT: Color = Color::rgba32(0x3a_3a_3a_ff);
+const BORDER_GREY: Color = Color::rgba32(0x5a_5a_5a_ff);
+const PRIMARY_LIGHT: Color = Color::rgba32(0x5c_c4_ff_ff);
+
 const TEXT_COLOR: Color = Color::rgb24(0xf0_f0_ea);
 const CURSOR_COLOR: Color = Color::WHITE;
 
 const BOX_HEIGHT: f64 = 24.;
-const BORDER_WIDTH: f64 = 2.;
+const FONT_SIZE: f64 = 14.0;
+const BORDER_WIDTH: f64 = 1.;
+const PADDING_TOP: f64 = 5.;
+const PADDING_LEFT: f64 = 4.;
 
+#[derive(Debug, Clone)]
 pub struct TextBox {
-    text: String,
     width: f64,
-    font: Option<<<Piet<'static> as RenderContext>::Text as Text>::Font>,
 }
 
 impl TextBox {
-    pub fn new(default_text: Option<String>, width: f64) -> TextBox {
-        TextBox {
-            text: default_text.unwrap_or_else(|| String::new()),
-            width,
-            font: None,
-        }
-    }
-    pub fn ui(self, ctx: &mut Ui) -> Id {
-        ctx.add(self, &[])
+    pub fn new(width: f64) -> TextBox {
+        TextBox { width }
     }
 
-    fn load_font(&mut self, rt: &mut Piet, font_size: f64) {
-        let font = rt
-            .text()
-            .new_font_by_name("Segoe UI", font_size)
+    fn get_layout(
+        &mut self,
+        text: &mut <Piet as RenderContext>::Text,
+        font_size: f64,
+        data: &String,
+    ) -> <Piet as RenderContext>::TextLayout {
+        // TODO: caching of both the format and the layout
+        let font = text
+            .new_font_by_name("Roboto", font_size)
             .unwrap()
             .build()
             .unwrap();
-
-        self.font = Some(font);
-    }
-
-    fn get_layout(&mut self, rt: &mut Piet, font_size: f64) -> <Piet as RenderContext>::TextLayout {
-        // TODO: caching of both the format and the layout
-        match &self.font {
-            Some(font) => {
-                return rt
-                    .text()
-                    .new_text_layout(&font, &self.text)
-                    .unwrap()
-                    .build()
-                    .unwrap()
-            }
-            _ => {
-                self.load_font(rt, font_size);
-
-                //QUESTION this recursion makes me uncomfortable
-                //but it solved my borrowing issues!
-                return self.get_layout(rt, font_size);
-            }
-        };
+        text.new_text_layout(&font, data).unwrap().build().unwrap()
     }
 }
 
-impl Widget for TextBox {
-    fn paint(&mut self, paint_ctx: &mut PaintCtx, geom: &Rect) {
-        let border_color = if paint_ctx.is_focused() {
-            ACTIVE_BORDER_COLOR
+impl Widget<String> for TextBox {
+    fn paint(
+        &mut self,
+        paint_ctx: &mut PaintCtx,
+        base_state: &BaseState,
+        data: &String,
+        _env: &Env,
+    ) {
+        let is_active = base_state.is_active();
```

Comment:
With the new focus stuff, I think this should be:

```rust
    let has_focus = base_state.has_focus();
```

(Now I'm slightly worried that I'm exposing inner details on the event flow by calling this "has_focus" instead of "is_focussed". Would love feedback on that)

Reply:
I don't know about inner details, but: I guess is_focussed is more consistent with is_hot / is_active. If you're relying on auto complete, if you type is_ you get all the relevant suggestions?